File: corpan/plugins/tauri-plugin-game-packs/android/src/main/java/com/corpora/tauri/gamepacks/GamePacksPlugin.kt
package com.corpora.tauri.gamepacks

import android.app.Activity
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import org.json.JSONArray

@InvokeArg
internal class PackRequest {
    lateinit var packId: String
}

@TauriPlugin
class GamePacksPlugin(private val activity: Activity) : Plugin(activity) {

    @Command
    fun listPacks(invoke: Invoke) {
        val packs = JSONArray()
        KNOWN_PACKS.forEach { info ->
            val obj = JSObject()
            obj.put("id", info.id)
            obj.put("name", info.name)
            obj.put("version", info.version)
            packs.put(obj)
        }
        val payload = JSObject()
        payload.put("packs", packs)
        invoke.resolve(payload)
    }

    @Command
    fun getManifestUrl(invoke: Invoke) {
        val args = try {
            invoke.parseArgs(PackRequest::class.java)
        } catch (e: Exception) {
            invoke.reject("Invalid args: ${e.message}")
            return
        }

        val packId = args.packId
        val payload = JSObject()
        payload.put("url", buildManifestUrl(packId))
        invoke.resolve(payload)
    }

    private fun buildManifestUrl(packId: String): String {
        return "http://corpan-pack.localhost/$packId/manifest.json"
    }

    companion object {
        private val KNOWN_PACKS = listOf(
            PackInfo("endless_learner", "Endless Learner", "0.1.0")
        )
    }
}

private data class PackInfo(
    val id: String,
    val name: String,
    val version: String?
)

File: corpan/plugins/tauri-plugin-game-packs/ios/Sources/GamePacksPlugin.swift
import Foundation
import Tauri

private struct PackRequest: Decodable {
    let packId: String
}

private struct PackInfo: Encodable {
    let id: String
    let name: String
    let version: String?
}

final class GamePacksPlugin: Plugin {
    private let knownPacks: [PackInfo] = []

    private var activeRequests: [String: NSBundleResourceRequest] = [:]

    @objc public func listPacks(_ invoke: Invoke) {
        invoke.resolve(knownPacks)
    }

    @objc public func getManifestUrl(_ invoke: Invoke) {
        do {
            let args = try invoke.parseArgs(PackRequest.self)
            ensurePack(packId: args.packId, invoke: invoke)
        } catch {
            invoke.reject("Invalid args: \(error)")
        }
    }

    private func ensurePack(packId: String, invoke: Invoke) {
        let request = NSBundleResourceRequest(tags: Set([packId]))
        activeRequests[packId] = request

        request.beginAccessingResources { [weak self] error in
            if let error = error {
                self?.activeRequests.removeValue(forKey: packId)
                invoke.reject("ODR error: \(error.localizedDescription)")
                return
            }

            do {
                let manifestUrl = try self?.installPack(packId: packId) ?? ""
                self?.activeRequests.removeValue(forKey: packId)
                request.endAccessingResources()
                if manifestUrl.isEmpty {
                    invoke.reject("Manifest url missing")
                } else {
                    invoke.resolve(manifestUrl)
                }
            } catch {
                self?.activeRequests.removeValue(forKey: packId)
                request.endAccessingResources()
                invoke.reject("Install error: \(error)")
            }
        }
    }

    private func installPack(packId: String) throws -> String {
        let fm = FileManager.default
        guard let supportDir = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else {
            throw NSError(domain: "corpan", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing app support dir"])
        }

        let destDir = supportDir
            .appendingPathComponent("corpan-packs", isDirectory: true)
            .appendingPathComponent(packId, isDirectory: true)

        if !fm.fileExists(atPath: destDir.path) {
            try fm.createDirectory(at: destDir, withIntermediateDirectories: true)
        }

        guard let bundleRoot = Bundle.main.resourceURL?
            .appendingPathComponent("corpan-packs")
            .appendingPathComponent(packId) else {
            throw NSError(domain: "corpan", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing bundle resources"])
        }

        try copyDirectory(from: bundleRoot, to: destDir)
        return buildManifestUrl(packId: packId)
    }

    private func buildManifestUrl(packId: String) -> String {
        return "corpan-pack://localhost/\(packId)/manifest.json"
    }

    private func copyDirectory(from source: URL, to destination: URL) throws {
        let fm = FileManager.default
        let resourceKeys: [URLResourceKey] = [.isDirectoryKey]
        let enumerator = fm.enumerator(at: source, includingPropertiesForKeys: resourceKeys)

        while let item = enumerator?.nextObject() as? URL {
            let relPath = item.path
                .replacingOccurrences(of: source.path, with: "")
                .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            let target = destination.appendingPathComponent(relPath)
            let values = try item.resourceValues(forKeys: Set(resourceKeys))

            if values.isDirectory == true {
                if !fm.fileExists(atPath: target.path) {
                    try fm.createDirectory(at: target, withIntermediateDirectories: true)
                }
            } else {
                if fm.fileExists(atPath: target.path) {
                    try fm.removeItem(at: target)
                }
                try fm.copyItem(at: item, to: target)
            }
        }
    }
}

@_cdecl("init_plugin_game_packs")
func init_plugin_game_packs() -> Plugin {
    return GamePacksPlugin()
}
